### PR TITLE
Add Emoji Support

### PR DESCRIPTION
**Pull request by greenbigfrog**
(does the same as #176  just that this PR got tested, etc.)

### CODE REVIEW

File: lib/discordrb/data.rb [hunk 1]
```diff
@@ -1568,12 +1581,127 @@ def from_bot?
       @author && @author.current_bot?
     end
 
+    # Check if any emoji got used in this message
+    # @return [boolean] wether any emoji got used or not
```
Comment (meew0) [inline lib/discordrb/data.rb]: "whether"

File: lib/discordrb/data.rb [hunk 2]
```diff
@@ -1568,12 +1581,127 @@ def from_bot?
...
+
+      @name = data['name']
+      @server = server
+      @id = data['id'].to_i
+
+      process_roles(data['roles']) if server
+    end
+
+    # @return [true, false] whether a server is attached to this emoji
+    def attached?
```
Comment (meew0) [inline lib/discordrb/data.rb]: Why does the attachment stuff still exist? I thought it has been replaced by `GlobalEmoji`/`Emoji`.

File: lib/discordrb/data.rb [hunk 3]
```diff
@@ -1568,12 +1581,127 @@ def from_bot?
...
+      !@server.nil?
+    end
+
+    # @return [String] the layout to mention it (or have it used) in a message
+    def mention
+      "<:#{@name}:#{@id}>"
+    end
+    alias_method :use, :mention
```
Comment (meew0) [inline lib/discordrb/data.rb]: Please add a newline between the method and the alias here.

File: lib/discordrb/data.rb [hunk 4]
```diff
@@ -1568,12 +1581,127 @@ def from_bot?
...
+      API.emoji_icon_url(@id)
+    end
+
+    # The inspect method is overwritten to give more useful output
+    def inspect
+      "<GlobalEmoji name=#{@name} id=#{@id}>"
+    end
+
+    # @!visibility private
+    def push_roles(emoji)
```
Comment (meew0) [inline lib/discordrb/data.rb]: This method doesn't seem to be used anywhere.

File: lib/discordrb/bot.rb [hunk 5]
```diff
@@ -382,13 +410,23 @@ def update_oauth_application(name, redirect_uris, description = '', icon = nil)
...
+      if /<@!?(?<id>\d+)>?/ =~ mention
+        user(id.to_i)
+      elsif /<@&(?<id>\d+)>?/ =~ mention
+        return nil unless server
+        server.role(id)
+      elsif /<:(\w+):(?<id>\d+)>?/ =~ mention
+        emoji.each do |element|
+          return element if element.id.to_i == id.to_i
+        end
+        return nil
```
Comment (meew0) [inline lib/discordrb/bot.rb]: This whole block inside the `elsif` with the `each` and the `return nil` can be replaced with `emoji.find { |element| ... }`

File: lib/discordrb/bot.rb [hunk 6]
```diff
@@ -145,6 +145,34 @@ def servers
...
+    # @overload emoji
+    #   The list of emoji the bot can use.
+    #   @return [Array<GlobalEmoji>] the emoji available.
+    def emoji(id = nil)
+      gateway_check
+      if id
+        @emoji = emoji
+        emoji = @emoji.find { |sth| sth.id == id }
+        return nil unless emoji
+        return emoji
```
Comment (meew0) [inline lib/discordrb/bot.rb]: The previous four lines can be simplified to just `emoji.find { |sth| ... }`.

File: lib/discordrb/api.rb [hunk 7]
```diff
@@ -107,6 +107,11 @@ def widget_url(server_id, style = 'shield')
     "#{api_base}/guilds/#{server_id}/widget.png?style=#{style}"
   end
 
+  # Make an emoji icon URL from emoji IDs
```
Comment (meew0) [inline lib/discordrb/api.rb]: "IDs" is wrong as only one ID is used.

File: lib/discordrb/data.rb [hunk 8]
```diff
@@ -1568,12 +1581,127 @@ def from_bot?
...
+      !@server.nil?
+    end
+
+    # @return [String] the layout to mention it (or have it used) in a message
+    def mention
+      "<:#{@name}:#{@id}>"
+    end
+    alias_method :use, :mention
+
+    # The icon URL of the emoji
```
Comment (meew0) [inline lib/discordrb/data.rb]: This should specify return type

File: lib/discordrb/data.rb [hunk 9]
```diff
@@ -1568,12 +1581,127 @@ def from_bot?
...
+      @id = data.id
+      @role_associations = Hash.new([])
+      @role_associations[data.server.id] = data.roles
+    end
+
+    # @return [String] the layout to mention it (or have it used) in a message
+    def mention
+      "<:#{@name}:#{@id}>"
+    end
+    alias_method :use, :mention
```
Comment (meew0) [inline lib/discordrb/data.rb]: Please add a newline in between the alias and the method.

File: lib/discordrb/data.rb [hunk 10]
```diff
@@ -1568,12 +1581,127 @@ def from_bot?
...
+
+    # The icon URL of the emoji
```
Comment (meew0) [inline lib/discordrb/data.rb]: This should also specify return type

File: lib/discordrb/bot.rb (hunk 6 above)
Comment (greenbigfrog) [inline lib/discordrb/bot.rb]: so just following code?
```ruby
emoji = @emoji.find { |sth| sth.id == id }
```

File: lib/discordrb/data.rb (hunk 2 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: Oops...

File: lib/discordrb/bot.rb [hunk 11]
```diff
@@ -145,6 +145,34 @@ def servers
       @servers
     end
 
+    # @overload emoji(id)
+    #   Return a emoji by its ID
```
Comment (greenbigfrog) [inline lib/discordrb/bot.rb]: fixed

File: lib/discordrb/bot.rb (hunk 6 above)
Comment (greenbigfrog) [inline lib/discordrb/bot.rb]: done

File: lib/discordrb/bot.rb [hunk 12]
```diff
@@ -382,13 +410,23 @@ def update_oauth_application(name, redirect_uris, description = '', icon = nil)
       API.update_oauth_application(@token, name, redirect_uris, description, icon)
     end
 
-    # Gets the user from a mention of the user.
-    # @param mention [String] The mention, which should look like <@12314873129>.
-    # @return [User] The user identified by the mention, or `nil` if none exists.
-    def parse_mention(mention)
+    # Gets the user or emoji from a mention of the user or emoji.
```
Comment (greenbigfrog) [inline lib/discordrb/bot.rb]: done

File: lib/discordrb/bot.rb [hunk 13]
```diff
@@ -382,13 +410,23 @@ def update_oauth_application(name, redirect_uris, description = '', icon = nil)
...
+    # @param mention [String] The mention, which should look like `<@12314873129>`, `<@&123456789>` or `<:Name:126328:>`.
+    # @param server [Server, nil] The server of the associated mention. Required for role and emoji parsing.
```
Comment (greenbigfrog) [inline lib/discordrb/bot.rb]: done

File: lib/discordrb/bot.rb (hunk 5 above)
Comment (greenbigfrog) [inline lib/discordrb/bot.rb]: done

File: lib/discordrb/cache.rb [hunk 14]
```diff
@@ -125,6 +125,16 @@ def private_channel(id)
       @private_channels[id] = channel
     end
 
+    # Finds a emoji by its name.
```
Comment (greenbigfrog) [inline lib/discordrb/cache.rb]: done

File: lib/discordrb/cache.rb [hunk 15]
```diff
@@ -125,6 +125,16 @@ def private_channel(id)
...
+    # @param name [String] The emoji name that should be resolved.
+    # @return [Emoji, nil] the emoji identified by the name, or `nil` if it couldn't be found.
+    def find_emoji(name)
+      LOGGER.out("Resolving emoji #{name}")
+      emoji.select do |element|
+        element.name == name
+      end.first
```
Comment (greenbigfrog) [inline lib/discordrb/cache.rb]: done

File: lib/discordrb/cache.rb [hunk 16]
```diff
@@ -125,6 +125,16 @@ def private_channel(id)
       @private_channels[id] = channel
     end
 
+    # Finds a emoji by its name.
+    # @param name [String] The emoji name that should be resolved.
+    # @return [Emoji, nil] the emoji identified by the name, or `nil` if it couldn't be found.
+    def find_emoji(name)
```
Comment (greenbigfrog) [inline lib/discordrb/cache.rb]: moved

File: lib/discordrb/data.rb (hunk 1 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: done

File: lib/discordrb/data.rb (hunk 2 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: removed

File: lib/discordrb/data.rb (hunk 3 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: done

File: lib/discordrb/data.rb (hunk 8 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: done

File: lib/discordrb/data.rb (hunk 9 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: done

File: lib/discordrb/data.rb (hunk 10 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: done

File: lib/discordrb/data.rb (hunk 4 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: oops... removed

File: lib/discordrb/api.rb (hunk 7 above)
Comment (greenbigfrog) [inline lib/discordrb/api.rb]: done

File: lib/discordrb/data.rb [hunk 17]
```diff
@@ -1501,6 +1504,16 @@ def initialize(data, bot)
       @edited = !@edited_timestamp.nil?
       @id = data['id'].to_i
 
+      @emoji = []
+
+      unless data['content'].nil?
```
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: done

File: lib/discordrb/data.rb [hunk 18]
```diff
@@ -1581,10 +1570,24 @@ def from_bot?
...
+      emoji = emoji.grep(/<:(?<name>\w+):(?<id>\d+)>?/)
+      emoji.each do |element|
+        @emoji << @bot.parse_mention(element)
+      end
+      @emoji
+    end
+
     # Check if any emoji got used in this message
-    # @return [boolean] wether any emoji got used or not
+    # @return [boolean] whether any emoji got used or not
```
Comment (Daniel-Worrall) [inline lib/discordrb/data.rb]: Should be "whether or not any emoji got used"

File: lib/discordrb/data.rb [hunk 19]
```diff
@@ -1581,10 +1570,24 @@ def from_bot?
...
     def emoji?
-      return true unless @emoji.empty?
+      emoji = @content.split
+      emoji = emoji.grep(/<:(?<name>\w+):(?<id>\d+)>?/)
```
Comment (Daniel-Worrall) [inline lib/discordrb/data.rb]: Same code as L1577-1578. Can/should be refactored

File: lib/discordrb/data.rb [hunk 20]
```diff
@@ -1568,12 +1570,133 @@ def from_bot?
...
   end
 
+  # Server emoji
+  class Emoji
+    include IDObject
+
+    # @return [String] the emoji name
+    attr_reader :name
+
+    # @return [Server, nil] the server attached to this emoji. `nil` if it is not attached
```
Comment (meew0) [inline lib/discordrb/data.rb]: This still says something about attachments

File: lib/discordrb/data.rb [hunk 21]
```diff
@@ -1568,12 +1570,133 @@ def from_bot?
...
+    attr_reader :server
+
+    # @return [Array<Role>, nil] roles this emoji is active for, `nil` if nit attached to a server
```
Comment (meew0) [inline lib/discordrb/data.rb]: Same as this. Note that the `nil` type annotation will also have to be removed

File: lib/discordrb/data.rb (hunk 18 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: done

File: lib/discordrb/data.rb (hunk 20 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: done

File: lib/discordrb/data.rb (hunk 21 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: done

File: lib/discordrb/data.rb [hunk 22]
```diff
@@ -1568,12 +1570,139 @@ def from_bot?
...
+
+      emoji = scan_for_emoji
+      emoji.each do |element|
+        @emoji << @bot.parse_mention(element)
+      end
+      @emoji
+    end
+
+    # Check if any emoji got used in this message
+    # @return [boolean] whether or not any emoji got used
```
Comment (meew0) [inline lib/discordrb/data.rb]: This type parameter should be `true, false`

Comment (greenbigfrog) [inline lib/discordrb/data.rb]: done

File: lib/discordrb/bot.rb [hunk 23]
```diff
@@ -417,15 +417,18 @@ def update_oauth_application(name, redirect_uris, description = '', icon = nil)
...
       elsif /<@&(?<id>\d+)>?/ =~ mention
-        return nil unless server
-        server.role(id)
+        return server.role(id) if server
+        servers.each do |element|
+          role = element.role(id)
+          return role unless role.nil?
+        end
       elsif /<:(\w+):(?<id>\d+)>?/ =~ mention
         emoji.find { |element| return element if element.id.to_i == id.to_i }
```
Comment (meew0) [inline lib/discordrb/bot.rb]: This can still be simplified, by dropping the `return element if`.

File: lib/discordrb/bot.rb [hunk 24]
```diff
@@ -417,15 +417,18 @@ def update_oauth_application(name, redirect_uris, description = '', icon = nil)
...
         return nil
```
Comment (meew0) [inline lib/discordrb/bot.rb]: This statement is unnecessary.

File: lib/discordrb/bot.rb (hunk 23 above)
Comment (greenbigfrog) [inline lib/discordrb/bot.rb]: done

File: lib/discordrb/bot.rb (hunk 24 above)
Comment (greenbigfrog) [inline lib/discordrb/bot.rb]: done

File: lib/discordrb/bot.rb [hunk 25]
```diff
@@ -145,6 +145,39 @@ def servers
...
+        emoji.values
+      end
+    end
+
+    alias_method :emojis, :emoji
+    alias_method :all_emoji, :emoji
+
+    # Finds an emoji by its name.
+    # @param name [String] The emoji name that should be resolved.
+    # @return [Emoji, nil] the emoji identified by the name, or `nil` if it couldn't be found.
```
Comment (SnazzyPine25) [inline lib/discordrb/bot.rb]: Wouldn't this return a GlobalEmoji?

Comment (greenbigfrog) [inline lib/discordrb/bot.rb]: nice catch!

File: lib/discordrb/data.rb [hunk 26]
```diff
@@ -1568,12 +1570,139 @@ def from_bot?
...
+    # @return [String] the emoji name
+    attr_reader :name
+
+    # @return [Server] the server of this emoji
+    attr_reader :server
+
+    # @return [Array<Role>] roles this emoji is active for
+    attr_reader :roles
+
+    def initialize(data, bot, server)
```
Comment (SnazzyPine25) [inline lib/discordrb/data.rb]: You should utilize `data['managed']` since it determines if the emoji is from Better TTV or not

**Comment by meew0** [comment]
Postponing this review for 3.1.0 as the milestone indicates since I really want to get on with 3.0.0, sorry!

File: lib/discordrb/data.rb (hunk 26 above)
Comment (meew0) [inline lib/discordrb/data.rb]: That's not what it means, managed means it's managed by any external service (can also be Twitch or Beam itself)

**Pull request merged**